### PR TITLE
[cc2538] avoid possible deadlock in otPlatRadioTransmit

### PR DESCRIPTION
**Pull request by sjlongland**
In situations where lots of traffic is present on the 6LoWPAN network, it is possible for the CC2538 to receive many frames whilst preparing to transmit.  This can be exacerbated by blocking I/O calls being executed within the OpenThread user application, causing incoming frames to be missed.

The result is the transceiver finite state machine winds up in the RX overflow state, and will refuse to enter normal receive mode until the RX buffers are flushed.  Similarly, it is possible that delays within the CC2538 cause a delay in writing transmit data, causing a TX underflow.  Both of these cause a deadlock as the CPU waits for a flag that never gets set.

We should instead detect these conditions, and try to handle them to prevent deadlocks.

### CODE REVIEW

**Comment by sjlongland** [comment]
Err whoops… few failed builds in that lot… let me fix that. :-)

File: examples/platforms/cc2538/radio.c [hunk 1]
```diff
@@ -340,25 +364,78 @@ otError otPlatRadioTransmit(otInstance *aInstance, otRadioFrame *aFrame)
...
+                }
+
+                // Flush the RX buffer
+                HWREG(RFCORE_SFR_RFST) = RFCORE_SFR_RFST_INSTR_FLUSHRX;
+                HWREG(RFCORE_SFR_RFST) = RFCORE_SFR_RFST_INSTR_FLUSHRX;
 
-        setChannel(aFrame->mChannel);
-        setTxPower(aFrame->mPower);
+                // Clear RXOVERF and RXABO flags
+                HWREG(RFCORE_SFR_RFERRF) &=
```
[the remark this answers is not in the document]
Comment (sjlongland) [reply examples/platforms/cc2538/radio.c]: I would have thought so, but then I didn't see the flags getting cleared, so decided to do it explicitly.  Perhaps more experimentation is in order to confirm.

File: examples/platforms/cc2538/radio.c [hunk 2]
```diff
@@ -340,25 +364,78 @@ otError otPlatRadioTransmit(otInstance *aInstance, otRadioFrame *aFrame)
...
+                // In this case, the state of our driver mis-matches our state.  So force
+                // matters by clearing our channel variable and calling setChannel.  This
+                // should bring our radio into the RX state, which should allow us to go
+                // into TX.
+                sChannel = 0;
+                setChannel(aFrame->mChannel);
+            }
+
+            // Check for a transmitter buffer underflow.
+            if (HWREG(RFCORE_SFR_RFERRF) & RFCORE_SFR_RFERRF_TXUNDERF)
```
[the remark this answers is not in the document]
Comment (sjlongland) [reply examples/platforms/cc2538/radio.c]: I do agree there, but thought I did witness that occurring on one occasion, and considering the fix is trivial, it made sense to test for and fix the error condition to remove any possibility of it happening.

**Comment by sjlongland** [comment]
A heads up, that last push was a re-base against master to resolve a conflict.

**Comment by sjlongland** [comment]
```
Broadcast message from root@travis-job-030895e1-cad7-4a39-ac34-d7b0886b255e
	(unknown) at 3:21 ...
The system is going down for power off NOW!
```

Heh… wonderful timing.

**Comment by sjlongland** [comment]
Okay, so having done these fixes, I have not seen a single deadlock out of the CC2538 radio driver.

I've done some further changes to support interrupt-driven reception, such that if a frame is received while code is busy elsewhere, it can potentially read the frame out of the CC2538 buffer to avoid the buffer overflow we've been seeing which helps in applications where the code is blocked doing synchronous operations (filesystem I/O for instance).  I'm not sure if it's worth merging that here as well, or whether that should be a separate pull request.

Thoughts?

**Comment by sjlongland** [comment]
I've just re-based the branch again to bring in some of the other fixes and get everything in sync with OpenThread master.

Last commit seems to inadvertently dragged some autotools changes in, I'm not sure why this stuff gets modified when I do a build.  I've hopefully rolled those changes back in this last force-push. :-)

**Comment by sjlongland** [comment]
Code has once again been re-based off `master`.  I believe I've addressed the review comments in earlier commits, but given how long it's been and the fact that I've had to re-base this code several times now, it might need reviewing again.

That, or is it just easier to close the pull request as a "won't fix"?  It is getting time consuming to re-base this branch every few months.

**Comment by jwhui** [comment]
@sjlongland, thanks again for contributing this.  Apologies for the (long) delay in processing this PR.

**Comment by sjlongland** [comment]
Many thanks. :-)

**Pull request merged**